File: src/game/quests/hints.ts
import type { GameDatabase } from '../data/types'
import type { PlayerSave } from '../save/types'
import { MAIN_MAP_ID } from '../world/constants'
import { getLocationMapId } from '../world/travel'
import { asQuestRows, getQuestProgress } from './quests'
import { questActiveStepObjectives } from './steps'

/** First Hint target on an active quest step, or the step's Visit target. */
export function questVisitHintLocationId(db: GameDatabase, save: PlayerSave): string | null {
  for (const quest of asQuestRows(db)) {
    const questId = quest['Quest ID']
    if (getQuestProgress(save, questId).status !== 'active') continue
    const step = questActiveStepObjectives(db, save, quest)
    if (!step) continue
    if (step.hintLocationIds.length > 0) return step.hintLocationIds[0]!
    if (step.visitLocationIds.length > 0) return step.visitLocationIds[0]!
  }
  return null
}

/** Map node to pulse on [browseMapId], or null. */
export function questHintNodeId(
  db: GameDatabase,
  save: PlayerSave,
  browseMapId: string,
): string | null {
  const hintId = questVisitHintLocationId(db, save)
  if (!hintId) return null
  const hint = db.Locations.find((row) => row['Location ID'] === hintId)
  if (!hint) return null
  if (getLocationMapId(hint) === browseMapId) return hintId

  let parentId = typeof hint['Parent Location ID'] === 'string' ? hint['Parent Location ID'] : null
  while (parentId) {
    const parent = db.Locations.find((row) => row['Location ID'] === parentId)
    if (!parent) break
    if (getLocationMapId(parent) === browseMapId) return parentId
    parentId =
      typeof parent['Parent Location ID'] === 'string' ? parent['Parent Location ID'] : null
  }

  if (browseMapId === MAIN_MAP_ID) {
    const hintMapId = getLocationMapId(hint)
    const gateway = db.Locations.find((row) => {
      if (getLocationMapId(row) !== MAIN_MAP_ID) return false
      if (!(row['Location Type'] ?? '').toLowerCase().includes('sub-map gateway')) return false
      return db.Locations.some(
        (child) =>
          child['Parent Location ID'] === row['Location ID'] && getLocationMapId(child) === hintMapId,
      )
    })
    return gateway?.['Location ID'] ?? null
  }
  return null
}

/** Pulse the location-screen world-map chip while a Visit hint is outstanding. */
export function questHintsWorldMapButton(db: GameDatabase, save: PlayerSave): boolean {
  const hintId = questVisitHintLocationId(db, save)
  return hintId != null && save.currentLocationId !== hintId
}

File: packages/ik_rules/lib/src/quests/hints.dart
import 'package:collection/collection.dart';
import 'package:ik_content/ik_content.dart';

import '../save/generated/save_models.dart';
import '../world/constants.dart';
import '../world/travel.dart';
import 'quests.dart';
import 'steps.dart';

/// First Hint target on an active quest step, or the step's Visit target.
String? questVisitHintLocationId(GameDatabase db, PlayerSave save) {
  for (final quest in asQuestRows(db)) {
    final questId = quest['Quest ID'];
    if (questId is! String) continue;
    if (getQuestProgress(save, questId).status != 'active') continue;
    final step = questActiveStepObjectives(db, save, quest);
    if (step == null) continue;
    if (step.hintLocationIds.isNotEmpty) return step.hintLocationIds.first;
    if (step.visitLocationIds.isNotEmpty) return step.visitLocationIds.first;
  }
  return null;
}

/// Map node to pulse on [browseMapId], or null.
String? questHintNodeId(GameDatabase db, PlayerSave save, String browseMapId) {
  final hintId = questVisitHintLocationId(db, save);
  if (hintId == null) return null;
  final hint = db.locations.where((row) => row.locationId == hintId).firstOrNull;
  if (hint == null) return null;
  if (getLocationMapId(hint) == browseMapId) return hintId;

  var parentId = hint.parentLocationId;
  while (parentId != null && parentId.isNotEmpty) {
    final parent = db.locations.where((row) => row.locationId == parentId).firstOrNull;
    if (parent == null) break;
    if (getLocationMapId(parent) == browseMapId) return parentId;
    parentId = parent.parentLocationId;
  }

  if (browseMapId == mainMapId) {
    final hintMapId = getLocationMapId(hint);
    for (final row in db.locations) {
      if (getLocationMapId(row) != mainMapId) continue;
      if (!(row.locationType ?? '').toLowerCase().contains('sub-map gateway')) continue;
      final opensHintMap = db.locations.any(
        (child) => child.parentLocationId == row.locationId && getLocationMapId(child) == hintMapId,
      );
      if (opensHintMap) return row.locationId;
    }
  }
  return null;
}

/// Pulse the location-screen world-map chip while a Visit hint is outstanding.
bool questHintsWorldMapButton(GameDatabase db, PlayerSave save) {
  final hintId = questVisitHintLocationId(db, save);
  return hintId != null && save.currentLocationId != hintId;
}
